src/meta/src/hummock/hummock_manager.rs

```diff
@@ -494,6 +479,36 @@ where
         ret
     }
 
+    #[named]
+    pub async fn unpin_version_before(
+        &self,
+        context_id: HummockContextId,
+        unpin_before: HummockVersionId,
+    ) -> Result<()> {
+        let mut versioning_guard = write_lock!(self, versioning).await;
+        let _timer = start_measure_real_process_timer!(self);
+        let versioning = versioning_guard.deref_mut();
+        let mut pinned_versions = VarTransaction::new(&mut versioning.pinned_versions);
+        let mut context_pinned_version = pinned_versions.new_entry_txn_or_default(
+            context_id,
+            HummockPinnedVersion {
+                context_id,
+                version_id: vec![0],
+            },
+        );
+
+        *context_pinned_version.version_id.first_mut().unwrap() = unpin_before;
+        commit_multi_var!(self, Some(context_id), context_pinned_version)?;
+
+        #[cfg(test)]
+        {
+            drop(versioning_guard);
+            self.check_state_consistency().await;
+        }
+
+        Ok(())
+    }
+
     #[named]
     pub async fn unpin_version(
```

Comment:
unpin_version will break the new assumption "version_pin.version_id is not empty". What about remove unpin_version, change version_pin.version_id to a single value and rename it to something like min_pinned_id?